### PR TITLE
Support plugin only request during plugin initialization

### PR DESCRIPTION
**Pull request by ben181231**
connect #219

### CODE REVIEW

**Comment by oursky-travis** [comment]
@ben181231, thanks for your PR! By analyzing the annotation information on this pull request, we identified @rickmak to be a potential reviewer

File: pkg/server/preprocessor/plugin.go [hunk 1]
```diff
@@ -26,11 +26,29 @@ type EnsurePluginReadyPreprocessor struct {
...
+		}
+
+		return http.StatusOK
+	}
+
+	response.Err = skyerr.NewError(
+		skyerr.PluginUnavailable,
+		"plugins are unavailable at the moment",
+	)
+	return http.StatusServiceUnavailable
```
Comment (rickmak) [inline pkg/server/preprocessor/plugin.go]: I think a test case for this preprocessor is needed as it before more complicated.

File: pkg/server/preprocessor/plugin.go [hunk 2]
```diff
@@ -26,11 +26,29 @@ type EnsurePluginReadyPreprocessor struct {
...
+	payload *router.Payload,
+	response *router.Response,
+) int {
+	if p.PluginContext.IsReady() {
+		return http.StatusOK
 	}
-	return http.StatusOK
+
+	if p.PluginContext.IsInitialized() {
+		if !payload.HasMasterKey() || !payload.IsFromPlugin() {
```
Comment (rickmak) [inline pkg/server/preprocessor/plugin.go]: The isFormPlugin() should only have effect if there is present of `master_key`.

Comment (ben181231) [inline pkg/server/preprocessor/plugin.go]: Okay. Maybe I should rename it to `IsPluginRequest()`

File: pkg/server/preprocessor/plugin.go (hunk 1 above)
Comment (ben181231) [inline pkg/server/preprocessor/plugin.go]: Agree.

**Comment by cheungpat** [comment]
I will leave more comments

File: pkg/server/skyerr/skyerr.go [hunk 3]
```diff
@@ -119,6 +119,9 @@ const (
 	// PluginTimeout occurs when an operation carried by a plugin is timed out
 	PluginTimeout
 
+	// PluginInitializing occurs when any of the plugins are initializing
```
Comment (cheungpat) [inline pkg/server/skyerr/skyerr.go]: “Error codes for expected error condition should be placed above this line.”

i.e. do not insert errors in the middle of a list

File: pkg/server/router/model.go [hunk 4]
```diff
@@ -150,6 +150,13 @@ func (p *Payload) HasMasterKey() bool {
 	return p.AccessKey == MasterAccessKey
 }
 
+// IsValidPluginRequest returns true when the payload indicate that the
+// request is from plugin AND the master key is used
+func (p Payload) IsValidPluginRequest() bool {
+	fromPlugin, _ := p.Data["_from_plugin"].(bool)
```
Comment (cheungpat) [inline pkg/server/router/model.go]: It is the job of a preprocessor to decide whether the request is a valid plugin request, not the `Request` struct. Move this function out of here.

File: pkg/server/plugin/handler.go [hunk 5]
```diff
@@ -56,12 +56,22 @@ func NewPluginHandler(info pluginHandlerInfo, ppreg router.PreprocessorRegistry,
...
+			"authenticator",
+			"dbconn",
+			"inject_user",
+			"require_user",
+			"plugin_ready",
+		)
 	} else if h.AccessKeyRequired {
 		h.preprocessors = h.PreprocessorList.GetByNames(
-			"plugin", "authenticator")
+			"authenticator",
```
Comment (cheungpat) [inline pkg/server/plugin/handler.go]: do you need `access_key` here? (and above)

File: pkg/server/router/model.go (hunk 4 above)
Comment (cheungpat) [inline pkg/server/router/model.go]: Especially that the “HasMasterKey” function only returns valid value after a auth preprocessor update it.

File: pkg/server/skyerr/skyerr.go (hunk 3 above)
Comment (ben181231) [inline pkg/server/skyerr/skyerr.go]: Oh. I see.

File: pkg/server/router/model.go (hunk 4 above)
Comment (ben181231) [inline pkg/server/router/model.go]: Does it conflict with [this comment](https://github.com/SkygearIO/skygear-server/pull/221#discussion_r87138247) ?

File: pkg/server/plugin/handler.go (hunk 5 above)
Comment (ben181231) [inline pkg/server/plugin/handler.go]: The `access_key` is needed since the plugin handler is setting `AccessKeyRequired` to be true.

Comment (cheungpat) [inline pkg/server/plugin/handler.go]: Ah I see now, either `accesskey` or `authenticator` is required. So `authenticator` fits the bill here.

**Pull request merged**